Client/Frontend/Browser/HomePanel/NTPDataSource.swift

```diff
@@ -0,0 +1,240 @@
+                Preferences.NewTabPage.selectedCustomTheme.value != nil {
+                return (theme.wallpapers, .withQRCode(refCode), .randomOrderAvoidDuplicates)
+            }
+            
+            if let sponsor = sponsor {
+                let attemptSponsored = Preferences.NewTabPage.backgroundSponsoredImages.value
+                    && backgroundRotationCounter == NTPDataSource.sponsorshipShowValue
+                    && !PrivateBrowsingManager.shared.isPrivateBrowsing
+                
+                return attemptSponsored ?
+                    (sponsor.wallpapers, .withBrandLogo(sponsor.logo), .sponsoredRotation)
+                    : (standardBackgrounds, .regular, .randomOrderAvoidDuplicates)
+            }
+            
+            return (standardBackgrounds, .regular, .randomOrderAvoidDuplicates)
+        }()
+        
+        if backgroundSet.isEmpty { return nil }
+        
+        // Choosing the actual index / item to use
+        let backgroundIndex = { () -> Int in
+            switch strategy {
+            case .sponsoredRotation:
+                defer {
+                    sponsoredBackgroundRotationIndex += 1
+                    // Force a max, and wrap back down if it is hit.
+                    sponsoredBackgroundRotationIndex %= backgroundSet.count
+                }
+                return sponsoredBackgroundRotationIndex
+            case .randomOrderAvoidDuplicates:
+                let availableRange = 0..<backgroundSet.count
+                // This takes all indeces and filters out ones that were shown recently
+                let availableBackgroundIndeces = availableRange.filter {
+                    !self.lastBackgroundChoices.contains($0)
+                }
+                // Chooses a new random index to use from the available indeces
+                // -1 will result in a `nil` return
+                let chosenIndex = availableBackgroundIndeces.randomElement() ?? -1
+                assert(chosenIndex >= 0, "NTP index was nil, this is terrible.")
+                assert(chosenIndex < backgroundSet.count, "NTP index is too large, BAD!")
+                
+                // This index is now added to 'past' tracking list to prevent duplicates
+                self.lastBackgroundChoices.append(chosenIndex)
+                // Trimming to fixed length to release older backgrounds
+                
+                self.lastBackgroundChoices = self.lastBackgroundChoices
+                    .suffix(min(backgroundSet.count - 1, NTPDataSource.numberOfDuplicateAvoidance))
+                return chosenIndex
+            }
+        }()
+        
+        // Force back to `0` if at end
+        backgroundRotationCounter %= NTPDataSource.sponsorshipShowRate
+        // Increment regardless, this is a counter, not an index, so smallest should be `1`
+        backgroundRotationCounter += 1
+        
+        return (backgroundSet[backgroundIndex], backgroundType)
```

Comment:
if `backgroundIndex` is -1 wouldn't this result in an `Index out of range error`?

Reply:
potentially yes, we can guard against it by using `safe` subscript, I'll fix it